### DISCUSSION TITLE
RM-LXC "unclean shutdown" nach Host Reboot auf Pi5B unter Raspberry-Pi OS

### DISCUSSION THREAD

**Discussion by Baxxy13** [OpenCCU/OpenCCU#2762] [🇩🇪 Fragen&Antworten] [answered]
Ich spiele gerade ein bisschen mit dem Pi5B und habe unter aktuellem Raspberry-Pi OS Lite RaspberryMatic als LXC mittels `install-lxc.sh` aufgesetzt.
Installation lief soweit problemlos durch und RM läuft nach erstem "rumspielen" unauffällig.
Was nicht geht ist die LED des, auf GPIO gestecktem, RPI-RF-MOD. Das ist für mich aber erstmal nebensächlich.

Was mich wurmt ist der "unclean shutdown" wenn ich den Pi5B per `reboot` neu starte bzw. per `poweroff` herunterfahre ohne den LXC vorher manuell zu stoppen.

Ich sehe beim shutdown/reboot `Job lxc.service/stop running (2s / 1min 30s)`.
Das ist aber nach 3-4s durch, also viel zu kurz um den LXC sauber herunterzufahren (`lxc-stop raspberrymatic` braucht ca. 20s)

Da ich mit googlen nicht weiterkomme frage ich mal hier nach Ideen dazu.

Mal so als Nebeninfo:
Der Pi5B liegt im Idle (mit RPI-RF-MOD auf GPIO und USB->SATA SSD als Bootmedium) bei ca. 5W

**Comment by Baxxy13** [comment]
Irgendwie mag ich den LXC. 😉 
Nach intensiverem googlen konnte ich das Problem lösen indem ich in der `/var/lib/lxc/raspberrymatic/config` einen Eintrag hinzufügte:
`lxc.signal.stop = SIGRTMIN+14`

Ob das jetzt 100%ig richtig ist weiß ich nicht, da ich nicht alle Optionen ausprobiert habe. Jedenfalls funktioniert's und der LXC wird bei Host shutdown/reboot sauber beendet.

> STOP SIGNAL
Allows one to specify signal name or number to forcibly shutdown the container. This option allows signal to be specified in kill(1) fashion, e.g. SIGKILL, SIGRTMIN+14, SIGRTMAX-10 or plain number. The default signal is SIGKILL.

Ggf. müssen auch noch HALT SIGNAL und REBOOT SIGNAL optimiert werden, aber da konnte ich erstmal keine Probleme feststellen.

**Reply by jens-maus** [reply]
Also solch ein Problem bzgl. stop signal hatten wir vor kurzem ja auch in den Docker Installationsn (siehe https://github.com/jens-maus/RaspberryMatic/issues/2717). Gelöst wurde es da durch hinzufügen von `SIGTERM` in das `Dockerfile` damit docker dann beim herunterfahren dieses nutzt (siehe https://github.com/jens-maus/RaspberryMatic/commit/dcacd98b88a92ee35829f511cacf453a0ebe926d).

Müsste man also in der Tat also mal genauer testen und rausfinden ob hier `SIGRTMIN+14` oder `SIGTERM` "richtiger" ist... und natürlich auch mal schauen was `lxc.signal.reboot` (das per default auf `SIGINT` steht) so hergibt....

**Reply by Baxxy13** [reply]
Also `SIGTERM` oder `SIGRTMIN+14` scheinen sich nicht viel zu nehmen.
Mit beiden ist der "lxc.service/stop" nach ca. 20s durch und es gibt keinen "unclean shutdown".



> auch mal schauen was `lxc.signal.reboot` (das per default auf `SIGINT` steht) so hergibt....

Wie mache ich das?

**Reply by jens-maus** [reply]
> Also `SIGTERM` oder `SIGRTMIN+14` scheinen sich nicht viel zu nehmen. Mit beiden ist der "lxc.service/stop" nach ca. 20s durch und es gibt keinen "unclean shutdown".

Du meinst du machst ein `reboot` des HOSTS und der fährt dann den container richtig runter und dann wird nach dem neustart des Hosts und dem wieder starten des LXC kein "unclean shutdown" gemeckert? Oder hast du das "poweroff" innerhalb des containers gemacht? Weil beides muss man getrennt beobachten...

> > auch mal schauen was `lxc.signal.reboot` (das per default auf `SIGINT` steht) so hergibt....
> 
> Wie mache ich das?

Es sollte möglich sein via `sudo lxc-stop -r <CONTAINER>` den container zu einem reboot zu bewegen. Schau mal ob das durchgeht und passt oder ob man da nen anderes signal als `SIGINT` dann braucht?!?!

**Reply by Baxxy13** [reply]
Es geht nur um `reboot/poweroff` vom Host wenn der LXC nicht vorher z.B. `per lxc-stop xxx` beendet wurde.
Hier wird aktuell der LXC "hart" beendet was zum "unclean-shutdown" innerhalb des LXC führt.

Die `lxc-xxx` Befehle im laufendem System funktionieren alle problemlos in der default Konfiguration.

Füge ich `lxc.signal.stop = SIGTERM` oder `lxc.signal.stop = SIGRTMIN+14` in die LXC.config geht auch ein Host  `reboot/poweroff` sauber durch (dauert ca. 20s bis der LXC gestoppt wurde).

**Reply by jens-maus** [reply, marked as answer]
Ok, inzwischen hab ich das `lxc.signal.stop = SIGTERM` zum `install-lxc.sh` skript hinzugefügt. Probier mal ob nun bei frischen Anlegen via `install-lxc.sh` alles korrekt angelegt wird und das `lxc.signal.stop` nun auf `SIGTERM` zeigt und ein reboot/poweroff korrekt geht.

**Reply by Baxxy13** [reply]
Funktioniert nun korrekt. Kein "unclean-shutdown" mehr nach Host reboot/poweroff.

Dann bereite ich später mal das nächste Thema vor... 😉
